prelude.dx

```diff
@@ -697,20 +704,31 @@ def evalpoly (_:VSpace v) ?=> (coefficients:n=>v) (x:Float) : v =
   -- Evaluate a polynomial at x.  Same as Numpy's polyval.
   fold zero \i c. coefficients.i + x .* c
 
+
+'Monoid
+
+interface Monoid a:Type where
+  mempty : a
+  mcombine : a -> a -> a  -- can't use `<>` just for parser reasons?
+
+(<>) : Monoid a ?=> a -> a -> a = mcombine
+
 'Length-erased lists
 
 data List a:Type =
   AsList n:Int foo:(Fin n => a)
 
-def (<>) (x:List a) (y:List a) : List a =
-  (AsList nx xs) = x
-  (AsList ny ys) = y
-  nz = nx + ny
-  AsList _ $ for i:(Fin nz).
-    i' = ordinal i
-    case i' < nx of
-      True  -> xs.(fromOrdinal _ i')
-      False -> ys.(fromOrdinal _ (i' - nx))
+instance monoidList : Monoid (List a) where
+  mempty = AsList _ []
+  mcombine = \x y.
+    (AsList nx xs) = x
+    (AsList ny ys) = y
+    nz = nx + ny
+    AsList _ $ for i:(Fin nz).
+      i' = ordinal i
+      case i' < nx of
+        True  -> xs.(fromOrdinal _ i')
+        False -> ys.(fromOrdinal _ (i' - nx))
```

Comment:
I agree, that would be nice. But it will make codegen trickier. We'll need a runtime representation of that arbitrary index set whereas currently we only need a runtime representation of the size, which is just an integer.